### PR TITLE
33 sending notifications

### LINKED ISSUE #33: Sending notifications/alerts to users
# Goal

Email or Desktop notifications via a desktop client

# Plan

1. Research using an email library
1. Look at using sockets API or similar to send data between Pi and PC
1.

# Additional information

Any other useful information (if appropriate)

### PR DESCRIPTION
**Pull request by C-Begley**
# Details

- Created client and server which passes messages to be displayed in Notify Send. Both only work on Linux devices

# Screenshots/Gif (if appropriate)
Screenshot from 2021-03-11 12-58-29

# Checklist

## Code
- [x] I have added some tests
- [x] I have run the tests locally to ensure they all pass by running `.scripts/build.sh -enable-testing`

## Documentation
- [x] I have updated documentation appropriately (GitHub Pages)
- [x] I have tested the updated GitHub Pages locally to ensure the changes to documentation render correctly (see this [Guidance](https://ese-peasy.github.io/PosturePerfection/guidance.html) page)

Close #33

### CODE REVIEW

File: src/CMakeLists.txt [hunk 1]
```diff
@@ -8,13 +8,22 @@ set(CMAKE_CXX_LINK_FLAGS "${CMAKE_CXX_LINK_FLAGS} -latomic")
...
 set(LIBSRC
   iir.cpp
   inference_core.cpp
   post_processor.cpp
   pre_processor.cpp
   posture_estimator.cpp)
 
+#add_subdirectory(notifications ENABLE_EXPORTS)
+#include_directories("${CMAKE_CURRENT_BINARY_DIR}/../../src/notifications")
+
```
Comment (Ashwin-MJ) [inline src/CMakeLists.txt]: Could we remove these if they're not needed?

File: src/notifications/CMakeLists.txt [hunk 2]
```diff
@@ -0,0 +1,20 @@
+cmake_minimum_required(VERSION 3.10)
+
+# set the project name
+project(Notification)
+
+set(NOTFYSRC
+  server.cpp
+  notify.cpp
+  client.cpp
+  )
```
Comment (Ashwin-MJ) [inline src/notifications/CMakeLists.txt:10]: Didn't you say we didn't need this (and line 15) in this CMakeLists file?

File: src/notifications/server.cpp [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+  std::string middle("\" -u critical --icon ");
+  std::string end(
+      "/docs/images/"
+      "posture-logo-no-text.png");
+  std::string out = start + buff + middle + cwd + end;
+  std::cout << out;
+  system(out.c_str());
+}
+// shutdown(pi_fd, SHUT_RDWR);
+// close(pi_fd);
```
Comment (Ashwin-MJ) [inline src/notifications/server.cpp]: Do we need these two lines?

**Review by Ashwin-MJ — CHANGES_REQUESTED** [review]
Just a few comments, and should we remove the binaries as well?
- `src/notifications/NotificationClient`
- `src/notifications/NotificationServer`

File: src/notifications/CMakeLists.txt (hunk 2 above)
Comment (C-Begley) [reply src/notifications/CMakeLists.txt:10]: This is the CMake for making your own server and client, independent from the main make file. That's there for manual testing as the unit tests aren't working with sockets

Comment (Ashwin-MJ) [reply src/notifications/CMakeLists.txt:10]: Yeah but yesterday didn't you say it worked without setting `NOTFYSRC` and you just needed the executables?

Comment (C-Begley) [reply src/notifications/CMakeLists.txt:10]: Not in src/notifications/cmake list, perhaps I said it for a different one?

Comment (C-Begley) [reply src/notifications/CMakeLists.txt:10]: The stuff we pair programmed didn't include this. This was something I made before that

Comment (Ashwin-MJ) [reply src/notifications/CMakeLists.txt:10]: No, you added the `NOTFYSRC` variable and the `add_library` section below when we did the pair programming. There's no point in doing this if you don't use it anywhere?

Discord to discuss?

Comment (Ashwin-MJ) [reply src/notifications/CMakeLists.txt:10]: Ah i think i understand the confusion now, you renamed from `LIBSRC` to `NOTFYSRC`. Even then, i don't think you need the `add_library` line

Comment (C-Begley) [reply src/notifications/CMakeLists.txt:10]: I'll hope on discord

**Review by Andrew-Ritchie — APPROVED** [review]
Look great, good work. I don't have any suggestions on improvements!

**Review by miklasr — APPROVED** [review]
Given the test completed on Friday I am happy to merge if we want to use this as planned: the Pi is a client and the user's desktop machine is the server, with the client sending messages to the server.

As we still had issues with the Pi operating as the server, I would be inclined to keep this unmerged if we are still working on making the Pi a server that could send to multiple clients.

**Pull request merged**